rfcs/0000-custom-health-checks/README.md

```diff
@@ -0,0 +1,330 @@
+# RFC-0000 Custom Health Checks for Kustomization using Common Expression Language(CEL)
+
+**Status:** provisional
+
+**Creation date:** 2024-01-05
+
+**Last update:** 2024-01-05
+
+## Summary
+
+This RFC proposes to support customization of the status readers in `Kustomizations`
+during the `healthCheck` phase for custom resources. The user will be able to declare
+the needed `conditions` in order to compute a custom resource status.
+In order to provide flexibility, we propose to use `CEL` expressions to declare 
+the expected conditions and their status.
+This will introduce a new field `customHealthChecks` in the `Kustomization` CRD
+which will be a list of `CustomHealthCheck` objects.
+
+## Motivation
+
+Flux uses the `Kstatus` library during the `healthCheck` phase to compute owned 
+resources status. This works just fine for all standard resources and custom resources
+that comply with `Kstatus` interfaces.
+
+In the current Kustomization implementation, we have addressed such a problem for
+kubernetes Jobs. We have implemented a `customJobStatusReader` that computes the
+status of a Job based on a defined set of conditions. This is a good solution for
+Jobs, but it is not generic and thus not applicable to other custom resources.
+
+Another use case is relying on non-standard `conditions` to compute the status of
+a custom resource. For example, we might want to compute the status of a custom
+resource based on a condtion other then `Ready`. This is the case for `Resources`
+that do intermediate patching like `Certificate` where you should look at the `Issued`
+condition to know if the certificate has been issued or not before looking at the
+`Ready` condition.
+
+In order to provide a generic solution for custom resources, that would not imply
+writing a custom status reader for each new custom resource, we need to provide a
+way for the user to express the `conditions` that need to be met in order to compute
+the status of a given custom resource. And we need to do this in a way that is
+flexible enough to cover all possible use cases, without having to change `Flux`
+source code for each new use case.
+
+### Goals
+
+- provide a generic solution for user to customize the health check of custom resources
+- support non-standard resources in `kustomize-controller`
+
+### Non-Goals
+
+- We do not plan to support custom `healthChecks` for core resources.
```

Comment:
I'm curious why this restriction ?

It seems perhaps arbitrary to not allow custom heathChecks for core resources. For instance for a Deployment, depending on the situation we might sometimes want ready to mean "all pods are ready", or "at least one pod is ready", or "at least minAvailable pods are ready".

Reply:
The reason for this is that kstatus covers all the bultin types, excluding them on a per kind bases is not possible, so we would need to drop them all and make users write hundreds of lines of CEL.

Reply:
The scope of this RFC is to allow defining health checks for custom resources. Once this is implemented, we could consider expanding CEL to native resources in a followup RFC, if it deems to be feasible and if Flux users will ask for it. So far, no one asked for more than what kstatus does for builtin kinds.

Reply:
To expand on Stefan answer, the default poller uses a slice of readers:

```go
  statusReaders = append(statusReaders,
    deploymentStatusReader,
    statefulSetStatusReader,
    replicaSetStatusReader,
    defaultStatusReader,
  )
```

There is a `Support()` method that will be used to find a reader that support a given objet `Kind`. The `defaultStatusReader` is the one that is used for all custom resources. It is able to compute the status if the CR is compliant with `Kstatus`. The actual proposal is to expand this list. So in order to support `native kinds` we would need to remove the other Readers.

Another thing to consider is the `SSA`. Objects are applied in stages, and we perform a normalization of all `native kinds` mainly working around various upstream Kubernetes API bugs. This implies that we expect a certain behaviour. Also normalization could invalidate a `CEL` expression as we pass the whole object to the evaluator.

Reply:
+1 to keeping the thing simple

my ideas of refinements may prove relevant later... or not ;)